### PR TITLE
Improve rejoining logic

### PR DESCRIPTION
**Pull request by CallumMoseley**
Readying up now generates a "ReadyId" on the server, which is unique for each time a player readies up (meaning it can identify both a player and a room). This is immediately saved on the client. If the player crashes or leaves before saving the file, they can create a new file, connect to MW, and press "Rejoin" to get the RandoResult that they missed (meaning the entire game won't need to be recreated).

### CODE REVIEW

File: MultiWorldServer/Server.cs [hunk 1]
```diff
@@ -121,9 +121,9 @@ public void ListReady()
             foreach (var kvp in ready)
             {
                 string playerString = "";
-                foreach (var kvp2 in kvp.Value)
+                foreach (var key in kvp.Value.Keys)
                 {
-                    playerString += $"{Clients[kvp2.Key].Nickname}, ";
+                    playerString += $"{Clients[key].Nickname}, ";
                 }
                 if (playerString != "") playerString = playerString.Substring(0, playerString.Length - 2);
```
Comment (Shadudev) [inline MultiWorldServer/Server.cs]: Perhaps we could utilize String.Join for that matter

File: MultiWorldServer/Server.cs [hunk 2]
```diff
@@ -473,19 +476,13 @@ private void HandleReadyMessage(Client sender, MWReadyMessage message)
...
                 if (!ready.ContainsKey(sender.Room))
                 {
-                    ready[sender.Room] = new Dictionary<ulong, RandoSettings>();
-                } 
-                
-                ready[sender.Room][sender.UID] = message.Settings;
+                    ready[sender.Room] = new Dictionary<ulong, (int, RandoSettings)>();
+                }
+
+                int readyId = (new Random()).Next();
```
Comment (Shadudev) [inline MultiWorldServer/Server.cs:479]: Low chances, but this isn't necessarily unique 🌵

File: MultiWorldServer/Server.cs [hunk 3]
```diff
@@ -552,29 +549,34 @@ private void HandleUnreadyMessage(Client sender, MWUnreadyMessage message)
...
 
             GameSessions[sender.Session.randoId].Save(sender.Session.playerId);
         }
 
+        private void HandleRejoinMessage(Client sender, MWRejoinMessage message)
+        {
+            if (!unsavedResults.ContainsKey(message.ReadyID))
+            {
+                Log($"Bad rejoin attempt (readyId = {message.ReadyID}, UID = {sender.UID})");
+                return;
```
Comment (Shadudev) [inline MultiWorldServer/Server.cs]: no need

File: MultiWorldServer/Server.cs [hunk 4]
```diff
@@ -601,15 +595,17 @@ private void HandleStartMessage(Client sender, MWStartMessage message)
...
                 nicknames.Add(sender.Nickname);
 
                 ready[room].Remove(sender.UID);
 
                 foreach (var kvp in ready[room])
                 {
-                    clients.Add(Clients[kvp.Key]);
-                    settings.Add(kvp.Value);
+                    clients.Add(Clients[kvp.Key]);
+                    readyIds.Add(kvp.Value.Item1);
```
Comment (Shadudev) [inline MultiWorldServer/Server.cs:607]: I think this adds the sender's readyId twice

File: RandomizerMod3.0/MultiWorld/ClientConnection.cs [hunk 5]
```diff
@@ -467,7 +477,11 @@ public void SendItem(string loc, string item, int playerId)
 
         public void NotifySave()
         {
-            SendMessage(new MWSaveMessage());
+            if (RandomizerMod.Instance.MWSettings.LastReadyID != -1)
+            {
+                SendMessage(new MWSaveMessage { ReadyID = RandomizerMod.Instance.MWSettings.LastReadyID });
+            }
+            RandomizerMod.Instance.MWSettings.LastReadyID = -1;
```
Comment (Shadudev) [inline RandomizerMod3.0/MultiWorld/ClientConnection.cs:481]: Is that on purpose?

Comment (CallumMoseley) [reply RandomizerMod3.0/MultiWorld/ClientConnection.cs:481]: ah good catch

File: MultiWorldServer/Server.cs (hunk 2 above)
Comment (CallumMoseley) [reply MultiWorldServer/Server.cs:479]: I think it's fine, i'm willing to take 1/2^32 odds of something going weird

Comment (Shadudev) [reply MultiWorldServer/Server.cs:479]: lol np, more than a few things need to go wrong for this to act as an issue

File: MultiWorldServer/Server.cs (hunk 1 above)
Comment (Shadudev) [reply MultiWorldServer/Server.cs]: Not urgent for this PR, it really doesn't do much difference

**Pull request merged**